### ISSUE TITLE
Create core config for data/config/cache directories.

### ISSUE THREAD

**Issue by adamcik** [mopidy/mopidy#843]
Instead of having each backend set these up we should have these setup in core, like we do for proxy settings, and then extensions should just use ``$dir/$ext_name/`` without having to do anything more. In the case of extensions which already define their own settings for this we should change the default something which refers to say ``$DATA_DIR/ext_name`` or just kill the settings outright.

**Comment by jodal** [comment]
I suggest adding something like the following, with these default values:

``` ini
[dir]
cache = $XDG_CACHE_DIR/mopidy
config = $XDG_CONFIG_DIR/mopidy
data = $XDG_DATA_DIR/mopidy
```

Should we extend `mopidy.utils.path.expand_path()` to take `config['dir']` as an argument and use it to replace e.g. `$CACHE_DIR`, `$CONFIG_DIR`, and `$DATA_DIR` in paths with the proper values from `config['dir']`? I suggest we keep the `$XDG_*` variable support, but promote our own vars for most uses as they point directly to the `mopidy` dir in each XDG dir.

Other relevant configs in Mopidy core:

- `local/data_dir`
  - Either change default to `$DATA_DIR/local`, or
  - Remove in favor of `os.path.join(config['dir']['data'], 'local')`
- `local/media_dir` -- Unaffected.
- `local/playlists_dir` -- Change default to `$DATA_DIR/local/playlists`.

Comments welcome!

**Comment by adamcik** [comment]
Sounds reasonable, that is as long as we don't end up with a circular resolution.

**Comment by adamcik** [comment]
Put this under a `core` schema perhaps?

**Comment by jodal** [comment]
Like this?

``` ini
[core]
cache_dir = $XDG_CACHE_DIR/mopidy
config_dir = $XDG_CONFIG_DIR/mopidy
data_dir = $XDG_DATA_DIR/mopidy
```

**Comment by adamcik** [comment]
Mhm, would also be where a `max_tracklist_length` setting would fit in eventually.

**Comment by adamcik** [comment]
As mentioned in https://github.com/tkem/mopidy-local-sqlite/issues/59#issuecomment-100457938 a reasonable way to expose these dirs to extensions would be to combine the handling of this and `get_or_create_dir` in a `Extension.get_data_dir()` class method. We just need to make sure that the initial config pass gets the core scheme in addition to logging.

**Comment by tkem** [comment]
Would it make sense to also add similar `get_config_dir` and `get_cache_dir` methods? Though I currently don't see a clear distinction between `data_dir` and `cache_dir` for mopidy-local-sqlite...

**Comment by adamcik** [comment]
Yes I would add all three with clear documentation with what each should be used for. For mopidy-local-sqlite data dir is probably all you need. A cache dir would be for things like caching album art that can be refetched / generated as one possible usage example.

**Comment by tkem** [comment]
Another question is how to handle hyphenated `ext_name`s, like `local-sqlite` or `podcast-itunes`, e.g whether these should be converted to subdirectories.

**Comment by tkem** [comment]
@adamcik: totally agree with using `cache_dir` for extracted album art, but in principle the sqlite database can also be recreated from `local scan`, so it's also just eime sort of "cached" information.
However, this should be up to individual extensions to figure out, and providing all three methods sounds good to me.
Another question: should we mak `cache_dir` and `data_dir` different directories by default, so extensions can for example wipe/remove `cache_dir` on refresh? Should we also provide extension method for "clearing" `cache_dir` and/or `data_dir`?

**Comment by adamcik** [comment]
http://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html is what these map to BTW, as for how they are to be used. So `cache_dir` normally ends up being `~/.cache/mopidy/...` and it should always be safe for me to just do `rm -rf ~/.cache`, so yes `data_dir` and `cache_dir` need to be distinct IMO.

**Comment by tkem** [comment]
This would really be awesome to have in v1.1, since it perfectly fits some of the things I have in mind for podcasts.

**Comment by tkem** [comment]
Q: Could the "config" dir contain files that are *not* parsed by Mopidy at startup? For example XML files in a domain-specific format?

**Comment by jodal** [comment]
Yet, it can. Mopidy will only care about `$config_dir/mopidy.conf` and any file referred to in `mopidy.conf`, like e.g. `logging.conf`.

**Comment by tkem** [comment]
@jodal: I was thinking about init scripts etc., which (AFAIK) used filesystem globs for generating a list of config files that was passed to Mopidy. Can't remember exactly, but use case would be an extension that gets/creates a "private" config dir and wants to put some XML file into that. So that shouldn't conflict with any config processing done by Mopidy core.

**Comment by jodal** [comment]
That is only the Debian package, which includes config from `/usr/share/mopidy/conf.d/`. That dir is only intended to be used by the Debian packages of other Mopidy extensions, so they can install default config for run-as-system-service that differs from run-as-user.

The only use of that today is Mopidy-Spotify is use it setup a different cache dir when it's a system service. With this feature implemented, and Mopidy-Spotify using `core/cache_dir` instead, the Mopidy-Spotify Debian package will no longer need to put any default config in `/usr/share/mopidy/conf.d/`.

**Comment by tkem** [comment]
Thanks @jodal, couldn't remember the details; so this shouldn't be an issue.

**Comment by jodal** [comment]
Fixed by #1232.
